bot.py

```diff
@@ -51,28 +51,29 @@ def setup(self):
         self.variables = {}
 
         filenames = []
-        if not hasattr(self.config, 'enable'):
-            for fn in os.listdir(os.path.join(os.getcwd(), 'modules')):
-                if fn.endswith('.py') and not fn.startswith('_'):
-                    filenames.append(os.path.join(home, 'modules', fn))
-        else:
-            for fn in self.config.enable:
-                filenames.append(os.path.join(home, 'modules', fn + '.py'))
-
-        if hasattr(self.config, 'extra'):
-            for fn in self.config.extra:
-                if os.path.isfile(fn):
-                    filenames.append(fn)
-                elif os.path.isdir(fn):
-                    for n in os.listdir(fn):
-                        if n.endswith('.py') and not n.startswith('_'):
-                            filenames.append(os.path.join(fn, n))
+
+        # Default module folder + extra folders
+        module_folders = [os.path.join(home, 'modules')]
+        module_folders.extend(getattr(self.config, 'extra', []))
+        
+        excluded = getattr(self.config, 'exclude', [])
+        enabled = getattr(self.config, 'enable', [])
+
+        for folder in module_folders:
+            if os.path.isfile(folder):
+                filenames.append(folder)
+            elif os.path.isdir(folder):
+                for fn in os.listdir(folder):
+                    if fn.endswith('.py') and not fn.startswith('_'):
+                        name = os.path.basename(fn)[:-3]
+                        # If whitelist is present only include whitelisted
+                        # Never include blacklisted items
+                        if name in enabled or not enabled and name not in excluded:
```

Comment:
I feel it's a lot more clearer than it was before.

The excluded and enabled variables aren't called "excluded_modules" and "enabled_modules" because it makes the lines really long, but if that's not a problem I can gladly rename them.

fn can probably be renamed to file, or filename.

Besides that I sadly can't think of a way to improve the code.